### PR TITLE
Introduce JDBCEncoder and JDBCDecoder

### LINKED ISSUE #231: LocalDateTime is read back incorrectly outside UTC
### Version

4.2.0-SNAPSHOT (commit 251c5bd169506c89296beed57174792dc65f81db)

### Context

I have a TIMESTAMP column in h2. When I write a LocalDateTime, it is read back adjusted for UTC, which is not what I expect. I am in UTC-2 (CET). If I write `2021-07-21T19:16:33`, it comes back as `2021-07-21T17:16:33`. If I had been in UTC-4, it would have come back as `2021-07-21T15:16:33`. When querying the database manually, I can see that the timestamp is stored exactly as I wanted to (`2021-07-21T19:16:33`). I need to read it unchanged, regardless of timezone.

### Do you have a reproducer?

The following test can be inserted into `io.vertx.jdbcclient.JDBCPoolCustomTypesTest` to reproduce. Note that it overrides the JVM's default timezone to CET in order to avoid the test incidentally passing for anyone in UTC.

```java
  @Test
  public void testLocalDateTime(TestContext should) {
    final Async test = should.async();
    String sql = "INSERT INTO t (ts) VALUES (?)";
    // To reproduce the issue, we have to change our
    // timezone to something other than UTC:
    TimeZone.setDefault(TimeZone.getTimeZone("CET"));
    LocalDateTime now = LocalDateTime.parse("2021-07-21T19:16:33.412751");

    client
      .preparedQuery(sql)
      .execute(Tuple.of(now))
      .onFailure(should::fail)
      .onSuccess(rows -> client
        .query("SELECT ts FROM t WHERE ts IS NOT NULL")
        .execute()
        .onFailure(should::fail)
        .onSuccess(rows2 -> {
          Row row = rows2.iterator().next();
          should.assertEquals(now, row.getLocalDateTime("TS"));
          test.complete();
        }));
  }
```

The test will fail like this: `java.lang.AssertionError: Not equals : 2021-07-21T19:16:33.412751 != 2021-07-21T17:16:33.412751`

### LINKED ISSUE #232: TIMESTAMP WITH TIME ZONE is not supported
### Version

4.2.0-SNAPSHOT (commit 251c5bd)

### Context

I have an h2 database with a column of type `TIMESTAMP WITH TIME ZONE`. I expect this to map to `java.time.OffsetDateTime`. Instead, calling `Row.getOffsetDateTime` results in a ClassCastException: "class java.lang.String cannot be cast to class java.time.OffsetDateTime".

Inserting an `OffsetDateTime` works well and I can see that it's persisted correctly.

When I call `Row.getValue`, I get back the timestamp as a String.

While I'm working with h2, I have also tested on hsqldb, and I see the identical behaviour there.

### Do you have a reproducer?

I wrote the following test, heavily based on `JDBCPoolCustomTypesTest`. Note that it requires the h2 dependency in the POM to be upgraded to 1.4.200, since the existing h2 dependency is very old (2016) and doesn't support TIMESTAMP WITH TIME ZONE.

```java
package io.vertx.jdbcclient;

import io.vertx.ext.unit.Async;
import io.vertx.ext.unit.TestContext;
import io.vertx.sqlclient.Row;
import io.vertx.sqlclient.Tuple;
import org.junit.Before;
import org.junit.Test;

import java.sql.Connection;
import java.sql.DriverManager;
import java.time.OffsetDateTime;
import java.util.ArrayList;
import java.util.List;

/**
 * Requires upgrading the h2 dependency to 1.4.200 since the current (very old) version doesn't support
 * TIMESTAMP WITH TIME ZONE.
 */
public class JDBCPoolTimestampTest extends ClientTestBase {

  private static final JDBCConnectOptions options = new JDBCConnectOptions()
    .setJdbcUrl("jdbc:h2:mem:test-" + JDBCPoolTimestampTest.class.getSimpleName() + ";DB_CLOSE_DELAY=-1");

  private static final List<String> SQL = new ArrayList<>();

  static {
    SQL.add("drop table if exists t");
    SQL.add("create table t (tstz TIMESTAMP WITH TIME ZONE)");
  }

  public static void resetDb() throws Exception {
    Connection conn = DriverManager.getConnection(options.getJdbcUrl());
    for (String sql : SQL) {
      conn.createStatement().execute(sql);
    }
  }

  @Before
  public void setUp() throws Exception {
    resetDb();
    super.setUp();
  }

  @Override
  protected JDBCConnectOptions connectOptions() {
    return options;
  }

  /**
   * Tests that writing an OffsetTimeDate is read back identically.
   * Currently fails with: "java.lang.ClassCastException: class java.lang.String cannot be cast to class java.time.OffsetDateTime"
   */
  @Test
  public void testOffsetDateTime(TestContext should) {
    final Async test = should.async();

    String sql = "INSERT INTO t (tstz) VALUES (?)";
    OffsetDateTime now = OffsetDateTime.now();

    client
      .preparedQuery(sql)
      .execute(Tuple.of(now))
      .onFailure(should::fail)
      .onSuccess(rows -> client
        .query("SELECT tstz from t")
        .execute()
        .onFailure(should::fail)
        .onSuccess(rows2 -> {
          Row row = rows2.iterator().next();
          should.assertEquals(row.getOffsetDateTime("TSTZ"), now);
          test.complete();
        }));
  }
}
```

### PR DESCRIPTION
**Pull request by zero88**
- Fix #231
- Fix #232
- Fix https://github.com/eclipse-vertx/vertx-sql-client/issues/918

### CODE REVIEW

**Comment by zero88** [comment]
cc @vietj 

Data type converter in Vertx JDBC (both `JDBCClient` and `JDBCPool`) is not fully correct.
The conversion need to take care about metadata and column type from `java.sql.Types` to "cast" correct data type.

I need to refactor this method below, it will be affect to some functions such as `JDBCSQLRowStream`/`JDBCCallable` 
then this PR is not yet ready.

https://github.com/vert-x3/vertx-jdbc-client/blob/45f4d7c4c0add79599257a9b319dc17929bb9b27/src/main/java/io/vertx/ext/jdbc/impl/actions/JDBCStatementHelper.java#L186-L187

**Comment by zero88** [comment]
@vietj @pmlopes Hi man, could you have quick look my solution? It is quite huge refactor but almost change internal API.
In short for correct data type depends on JDBC driver, my solution introduces 2 SPI classes:
- `JDBCEncoder`: for convert user input to database value
- `JDBCDecoder`: for convert database value to java value

The default implementation is best efforts converter that I already tested on H2/HSQL/PostgreSQL but developer can customize them depends on JDBC driver.

If you agree with my solution, then I will continue implementing `JDBCEncoder` (`JDBCDecoder` is already done) and add more test.

**Comment by zero88** [comment]
@pmlopes @vietj pls give me a little feedback on my solution then I can continue

>In short for correct data type depends on JDBC driver, my solution introduces 2 SPI classes:
> JDBCEncoder: for convert user input to database value
> JDBCDecoder: for convert database value to java value

File: src/main/java/io/vertx/ext/jdbc/spi/JDBCDecoder.java [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+ * You can replace it to adapt to an specific SQL driver by creating your owns then includes in the SPI file
+ * ({@code META-INF/services/io.vertx.ext.jdbc.spi.JDBCDecoder})
+ *
+ * @see <a href=https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/jdbc_42.html">Mapping of java.sql.Types to SQL types</a>
+ * @see io.vertx.ext.jdbc.spi.impl.JDBCDecoderImpl
+ * @see java.sql.JDBCType
+ * @see java.sql.SQLType
+ * @since 4.2.0
+ */
+public interface JDBCDecoder extends JDBCParser {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/jdbc/spi/JDBCDecoder.java:24]: can you avoid extends here ? it's confusing

File: src/main/java/io/vertx/ext/jdbc/spi/JDBCEncoder.java [hunk 2]
```diff
@@ -0,0 +1,26 @@
...
+ * You can replace it to adapt to an specific SQL driver by creating your owns then includes in the SPI file
+ * ({@code META-INF/services/io.vertx.ext.jdbc.spi.JDBCEncoder})
+ *
+ * @see <a href=https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/jdbc_42.html">Mapping of java.sql.Types to SQL types</a>
+ * @see io.vertx.ext.jdbc.spi.impl.JDBCEncoderImpl
+ * @see java.sql.JDBCType
+ * @see java.sql.SQLType
+ * @since 4.2.0
+ */
+public interface JDBCEncoder extends JDBCParser {
```
Comment (vietj) [inline src/main/java/io/vertx/ext/jdbc/spi/JDBCEncoder.java:20]: can you avoid extends here ? it's confusing

**Comment by vietj** [comment]
_The default implementation is best efforts converter that I already tested on H2/HSQL/PostgreSQL but developer can customize them depends on JDBC driver._

can you elaborate ?

I'm in favor of this refactor

**Comment by zero88** [comment]
Ping @vietj 
could you or someone in team can review it?

**Comment by zero88** [comment]
@vietj  Already rebase as latest master
The [failed test](https://github.com/vert-x3/vertx-jdbc-client/runs/3337484187?check_suite_focus=true) has 2 tests that doesn't seem to relate my patch.

> Tests in error: 
>  PoolTest.testUseAvailableResources:71->AsyncTestBase.await:121->AsyncTestBase.await:130 » TestTimedOut
> ConcurrencyTest.concurrentQueriesOnSameContext » JdbcSQLNonTransientConnection

And 1 case relates to current implementation.

> ClickHouseOldAPITest.simpleTest » SQLFeatureNotSupported

```
Tests run: 1, Failures: 0, Errors: 1, Skipped: 0, Time elapsed: 9.467 sec <<< FAILURE! - in io.vertx.it.ClickHouseOldAPITest
simpleTest(io.vertx.it.ClickHouseOldAPITest)  Time elapsed: 9.467 sec  <<< ERROR!
java.sql.SQLFeatureNotSupportedException: null
	at ru.yandex.clickhouse.ClickHousePreparedStatementImpl.getParameterMetaData(ClickHousePreparedStatementImpl.java:442)
	at com.mchange.v2.c3p0.impl.NewProxyPreparedStatement.getParameterMetaData(NewProxyPreparedStatement.java:1159)
	at io.vertx.ext.jdbc.impl.actions.AbstractJDBCAction.fillStatement(AbstractJDBCAction.java:106)
	at io.vertx.ext.jdbc.impl.actions.JDBCQuery.execute(JDBCQuery.java:47)
	at io.vertx.ext.jdbc.impl.actions.JDBCQuery.execute(JDBCQuery.java:30)
	at io.vertx.ext.jdbc.impl.JDBCConnectionImpl.lambda$schedule$3(JDBCConnectionImpl.java:219)
	at io.vertx.core.impl.ContextImpl.lambda$null$0(ContextImpl.java:159)
	at io.vertx.core.impl.AbstractContext.dispatch(AbstractContext.java:100)
	at io.vertx.core.impl.ContextImpl.lambda$executeBlocking$1(ContextImpl.java:157)
	at io.vertx.core.impl.TaskQueue.run(TaskQueue.java:76)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at io.netty.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
	at java.base/java.lang.Thread.run(Thread.java:829)
```

`ClickHouse` is special jdbc driver that doesn't support [ParameterMetaData](https://docs.oracle.com/javase/8/docs/api/java/sql/PreparedStatement.html#getParameterMetaData--)

I will update API, and warn in log if any special jdbc has same behavior. Then a developer must implement it by himself.
How your thought?

**Comment by zero88** [comment]
@vietj Already fixed with ClickHouseJDBC

**Review by pmlopes — APPROVED** [review]

**Comment by pmlopes** [comment]
@zero88 I'm trying to merge this, but I see that tests are failing (locally)

```
Results :

Tests in error: 
  PoolTest.testUseAvailableResources:71->AsyncTestBase.await:121->AsyncTestBase.await:130 » TestTimedOut
  ConcurrencyTest.concurrentQueriesOnSameContext » JdbcSQLNonTransientConnection

Tests run: 144, Failures: 0, Errors: 2, Skipped: 4
```

If you manage to figure out why, I think I'm ready to merge it

File: src/main/java/io/vertx/ext/jdbc/spi/JDBCDecoder.java [hunk 3]
```diff
@@ -0,0 +1,76 @@
...
+
+/**
+ * Represents for JDBC decoder from SQL value to Java value
+ * <p>
+ * The default decoder provides the best efforts to convert {@code SQL type} to standard {@code Java type} as {@code JDBC 4.2} spec.
+ * <p>
+ * You can replace it to adapt to a specific SQL driver by creating your owns then includes in the SPI file
+ * ({@code META-INF/services/io.vertx.ext.jdbc.spi.JDBCDecoder})
+ *
+ * @see <a href=https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/jdbc_42.html">Mapping of java.sql.Types to SQL types</a>
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/jdbc/spi/JDBCDecoder.java:19]: The final `"` at the end of the url is breaking the syntax highlighter and will also create a broken link reference in the javadoc.

File: src/main/java/io/vertx/ext/jdbc/spi/JDBCEncoder.java [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+
+/**
+ * Represents for JDBC encoder from Java value to SQL value
+ * <p>
+ * The default encoder provides the best efforts to convert {@code Java type} to {@code SQL type} as {@code JDBC 4.2} spec.
+ * <p>
+ * You can replace it to adapt to a specific SQL driver by creating your owns then includes in the SPI file
+ * ({@code META-INF/services/io.vertx.ext.jdbc.spi.JDBCEncoder})
+ *
+ * @see <a href=https://docs.oracle.com/javase/8/docs/technotes/guides/jdbc/jdbc_42.html">Mapping of java.sql.Types to SQL types</a>
```
Comment (pmlopes) [inline src/main/java/io/vertx/ext/jdbc/spi/JDBCEncoder.java:16]: Broken url `"` at the end.

**Comment by zero88** [comment]
@pmlopes 
These test issues I think it is due to we upgrade to latest H2 version.

> ConcurrencyTest.concurrentQueriesOnSameContext 

`H2 v1.4.200` remove `MULTI_THREADED` property: [h2 github](https://github.com/h2database/h2database/issues/2150) and [changelog](http://www.h2database.com/html/changelog.html?highlight=MULTI_THREADED&search=MULTI_THREADED#firstFound). So should disable a test?

About this one:
> PoolTest.testUseAvailableResources:71

I'm not sure what thing is testing. Let me study a little bit

**Comment by zero88** [comment]
@pmlopes Fixed all comment. Could you pls recheck if it is ok

**Pull request merged**